### ISSUE TITLE
`Self` as return type annotation in base class method breaks multi-inheritance

### ISSUE THREAD

**Issue by uriyyo** [python/mypy#14640]
**Bug Report**

`Self` as return type annotation in base class method breaks multi-inheritance.

(A clear and concise description of what the bug is.)

**To Reproduce**

```python
import copy

from typing import Self


class Cloneable:
    def clone(self) -> Self:
        return copy.copy(self)


class Immutable:
    def clone(self) -> Self:
        return self


class Collection(Cloneable):
    pass


class Tuple(Immutable, Collection):
    pass
```

**Expected Behavior**
No errors

**Actual Behavior**
```py
error: Definition of "clone" in base class "Immutable" is incompatible with definition in base class "Cloneable"  [misc]
```

Same code using old approach woks as expected:
```py
import copy

from typing import TypeVar


_Self = TypeVar("_Self")


class Cloneable:
    def clone(self: _Self) -> _Self:
        return copy.copy(self)


class Immutable:
    def clone(self: _Self) -> _Self:
        return self


class Collection(Cloneable):
    pass


class Tuple(Immutable, Collection):
    pass
```

`pyright` doesn't raise an error.

**Your Environment**

- Mypy version used: 1.0.0
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.11

**Comment by hauntsaninja** [comment]
I guess the issue is that the bounds are different?

**Comment by uriyyo** [comment]
@hauntsaninja  I guess code with `Self` should cover all the cases when `TypeVar` workaround was used.

So if this code works as expected:
```py
import copy

from typing import TypeVar


_Self = TypeVar("_Self")


class Cloneable:
    def clone(self: _Self) -> _Self:
        return copy.copy(self)


class Immutable:
    def clone(self: _Self) -> _Self:
        return self


class Collection(Cloneable):
    pass


class Tuple(Immutable, Collection):
    pass
```

Then the code below should work too:
```py
import copy

from typing import Self


class Cloneable:
    def clone(self) -> Self:
        return copy.copy(self)


class Immutable:
    def clone(self) -> Self:
        return self


class Collection(Cloneable):
    pass


class Tuple(Immutable, Collection):
    pass
```

I didn't find anything related to this in [PEP 673](https://peps.python.org/pep-0673/), but `pyright` handle such case and works as I expected.

**Comment by srittau** [comment]
For the record, the `Collection` class is not necessary to expose the problem:

```python
from typing import Self

class Base1:
    def foo(self) -> Self:
        return self

class Base2:
    def foo(self) -> Self:
        return self

class Sub(Base1, Base2):  # error: Definition of "foo" in base class "Base1" is incompatible with definition in base class "Base2"  [misc]
    pass
```

(mypy 1.11.2)

Real-life example: python/typeshed#12788, where peewee makes use of that pattern.

**Comment by sterliakov** [comment]
@hauntsaninja this is one more ticket resolved by #18465 - sorry again, I should have searched for them more thoroughly:(

**Comment by hauntsaninja** [comment]
No need to apologise, thank you for fixing so many things!
